### ISSUE TITLE
Recent update breaks the plugin

### ISSUE THREAD

**Issue by mandeep** [SublimeLinter/SublimeLinter-flake8#56]
The latest update breaks sublimelinter-flake8. To confirm the update is the problem I reverted back to the v2.1.0 where everything works fine.

```
SublimeLinter: find_python(version=None, script='flake8', module=None) 
SublimeLinter: find_python: default python = /usr/bin/python 
SublimeLinter: find_python: available version = {'minor': 7, 'major': 2} 
SublimeLinter: find_python: 'flake8' path = /usr/bin/flake8 
SublimeLinter: find_python: <= ('/usr/bin/python', '/usr/bin/flake8', 2, 7) 
SublimeLinter: flake8 version query: /usr/bin/python /usr/bin/flake8 --version 
SublimeLinter: WARNING: no flake8 version could be extracted from:
Traceback (most recent call last):
  File "/usr/bin/flake8", line 6, in <module>
    from pkg_resources import load_entry_point
  File "/usr/lib/python2.7/dist-packages/pkg_resources/__init__.py", line 3019, in <module>
    @_call_aside
  File "/usr/lib/python2.7/dist-packages/pkg_resources/__init__.py", line 3003, in _call_aside
    f(*args, **kwargs)
  File "/usr/lib/python2.7/dist-packages/pkg_resources/__init__.py", line 3032, in _initialize_master_working_set
    working_set = WorkingSet._build_master()
  File "/usr/lib/python2.7/dist-packages/pkg_resources/__init__.py", line 655, in _build_master
    ws.require(__requires__)
  File "/usr/lib/python2.7/dist-packages/pkg_resources/__init__.py", line 963, in require
    needed = self.resolve(parse_requirements(requirements))
  File "/usr/lib/python2.7/dist-packages/pkg_resources/__init__.py", line 849, in resolve
    raise DistributionNotFound(req, requirers)
pkg_resources.DistributionNotFound: The 'flake8==3.2.1' distribution was not found and is required by the application
```

Here's is the output from flake8
```
mandeep@debian:~$ which flake8
/usr/bin/flake8
mandeep@debian:~$ /usr/bin/flake8 --version
3.2.1 (mccabe: 0.5.3, pycodestyle: 2.2.0, pyflakes: 1.3.0) CPython 3.5.3 on Linux
```

**Comment by joeycarter** [comment]
I'm also seeing this behaviour. In my case I get,

```
SublimeLinter: find_python(version=None, script='flake8', module=None) 
SublimeLinter: find_python: default python = /usr/bin/python 
SublimeLinter: find_python: available version = {'major': 2, 'minor': 7} 
SublimeLinter: find_python: 'flake8' path = /usr/local/bin/flake8 
SublimeLinter: find_python: <= ('/usr/bin/python', '/usr/local/bin/flake8', 2, 7) 
SublimeLinter: flake8 version query: /usr/bin/python /usr/local/bin/flake8 --version 
SublimeLinter: WARNING: no flake8 version could be extracted from:
Traceback (most recent call last):
  File "/usr/local/bin/flake8", line 7, in <module>
    from flake8.main.cli import main
ImportError: No module named flake8.main.cli
```

My crude workaround is to replace,

```python
cmd = ('flake8@python', '*', '-')
```

with,

```python
cmd = ('flake8', '*', '-')
```

in `linter.py`. Then SublimeLinter-flake8 uses the correct version of Python (`python3` in my case, since I installed `flake8` with `pip3`).

**Comment by mandeep** [comment]
Seems to be caused by #51

**Comment by JaniKarh** [comment]
An example of this issue. I have Python 3.6 installed via Homebrew, and pip-installed flake8 on system level. Sublime major version is 3 and build is 3152.

```
which python3
/usr/local/bin/python3

which flake8
/usr/local/bin/flake8
```
Flake8 works perfectly fine from shell.

Trying to load SublimeLinter results, with debug on:
```
SublimeLinter: find_posix_python: python3 => /usr/local/bin/python3 
SublimeLinter: find_python: system python = /usr/local/bin/python3 
SublimeLinter: find_python: available version = {'major': 3, 'minor': 6} 
SublimeLinter: find_python: <= ('/usr/local/bin/python3', None, 3, 6) 
SublimeLinter: sys.path for /usr/local/bin/python3:
/usr/local/Cellar/python3/3.6.3/Frameworks/Python.framework/Versions/3.6/lib/python36.zip
/usr/local/Cellar/python3/3.6.3/Frameworks/Python.framework/Versions/3.6/lib/python3.6
/usr/local/Cellar/python3/3.6.3/Frameworks/Python.framework/Versions/3.6/lib/python3.6/lib-dynload
/usr/local/lib/python3.6/site-packages
```
... and a couple of lines later:
```
SublimeLinter: find_python(version=None, script='flake8', module=None) 
SublimeLinter: find_python: default python = /usr/bin/python 
SublimeLinter: find_python: available version = {'major': 2, 'minor': 7} 
SublimeLinter: find_python: 'flake8' path = /usr/local/bin/flake8 
SublimeLinter: find_python: <= ('/usr/bin/python', '/usr/local/bin/flake8', 2, 7) 
SublimeLinter: flake8 version query: /usr/bin/python /usr/local/bin/flake8 --version 
SublimeLinter: WARNING: no flake8 version could be extracted from:
Traceback (most recent call last):
  File "/usr/local/bin/flake8", line 7, in <module>
    from flake8.main.cli import main
ImportError: No module named flake8.main.cli
```
 No matter the linter settings, everything seems to default to this `default python`in the end. I hope this helps to resolve the issue.

**Comment by jaraco** [comment]
This issue affects me also.

I decided to use the [git clone technique](https://packagecontrol.io/docs/customizing_packages#Git-Hg_Clone) to install the old version.

1. Using Package Control, remove the SublimeLinter-flake8 package.
2. In the SublimeText app data directory ([look here](https://www.sublimetext.com/docs/3/revert.html)), in `Packages/`: `git clone https://github.com/SublimeLinter/SublimeLinter-flake8 -b 2.1.0`.

**Comment by JaniKarh** [comment]
@jaraco I edited the package manually the way @joeycarter suggests above. Everything works again and I can continue my actual work.

With my understanding there might be something wrong with the way SublimeLinter resolves the @python executable path, regardless of the linter settings.

**Comment by asfaltboy** [comment]
@JaniKarh thanks for the info! According to your comments, you have flake8 installed in your python3 env, which is *not your default python* (at least /usr/bin/python seems to be 2.7). Executing python uses the python executable only, not python2 or python3, so whatever version it is (normally a symlink), is the env that will be used.
Do you have a `/usr/local/bin/python`? If so, what does it point to?

Hmm ... it seems that (at the very least) the resolution order of python versions differs when we log `SublimeLinter: find_python: default python = /usr/bin/python` vs the initial `SublimeLinter: find_python: system python = /usr/local/bin/python3`. This may be the real core of your issue. Could it be that you use an plugin that modifies environment, such as [Environment Settings][1] ? If so, this plugin it could be modifying paths, placing `/usr/bin/` before `/usr/local/bin` which is what you intended, because remember, the first python found is used to execute `flake8 --version`.

Also, I haven't followed the code for using "executable" setting so for simplicity sake, I'd remove that.

P.s.: I'm not affiliate or even a contributor to SublimeLinter, I just feel that I should help everyone out since I'm a fan of this PR and feel responsible for bumping it up into being merged.

[1]: https://packagecontrol.io/packages/Environment%20Settings

**Comment by asfaltboy** [comment]
@mandeep thanks for the detailed report, can you also attach your sublimelinter `settings` and related `project-settings` or `.sublimelinterrc` contents? I would put my money that in your case, the environment in sublime differs from your command line. I used the following [when debugging the issue reproduced locally][1].

```diff
 lint/util.py | 9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/lint/util.py b/lint/util.py
index bdd6613..b75beb0 100644
--- a/lint/util.py
+++ b/lint/util.py
@@ -1278,6 +1278,15 @@ def popen(cmd, stdout=None, stderr=None, output_stream=STREAM_BOTH, env=None, ex
         env.update(extra_env)
 
     try:
+        from . import persist
+        persist.debug('{} {} {} {} {} {}'.format(
+            cmd,
+            subprocess.PIPE,
+            stdout,
+            stderr,
+            info,
+            env
+        ))
         return subprocess.Popen(
             cmd,
             stdin=subprocess.PIPE,
```

[1]: https://github.com/SublimeLinter/SublimeLinter-flake8/pull/51#issuecomment-339053942

**Comment by jaraco** [comment]
On my system, the default `python` is also `python2.7`, in which `flake8` is not installed. What's awkward is that with this latest patch, SublimeLinter.flake8 is expecting `flake8` to be installed in that default python installation. That seems to be a mistake, especially given that the shebang for `flake8` clearly points to another Python.

```
bash-3.2$ which python
/usr/bin/python
bash-3.2$ head -n 1 $(which flake8)
#!/Library/Frameworks/Python.framework/Versions/3.6/bin/python3
```

The mistake I think is that this plugin is requiring the environment to be massaged to ensure that `python` is always the desired target executable, a change which could potentially affect many other applications.

Prior to #51, however, everything seemed to work fine. Contrary to the claim in that ticket:

> Unless cmd is defined with @python, SublimeLinter will always use ST's
python, which is version 3.

That can't be the case, because with 2.1.0, the `flake8` processing happens and happens with Python 3.6 syntax checks (e.g. recognizing "f-strings"), even though the ST python is Python 3.3 (or similar).

In my opinion, the previous behavior was correct and honored the behavior one would find with the command line instead of explicitly selecting a (possibly wrong) interpreter.

**Comment by JaniKarh** [comment]
@asfaltboy you are correct, the default Python in `/usr/bin/python` is indeed 2.7, and is the macOS provided Python, which I do not use at all (per se).

And there is nothing in `/usr/local/bin/python`. Only python3 exists. You may have pointed me to right direction, I will try to symlink a bit later.

And last, but not least, thank you very much for your support!

**Comment by asfaltboy** [comment]
@jaraco regarding the shebang, that's a very correct observation, however I don't believe SublimeLinter is "shebang aware"; that is, it does not parse the linter script to figure out the right python version. Maybe that would be a nice solution, at least for "POSIX systems".

As far as I know, the version of python used to run flake8 is used to "enforce" syntax for a given version. I.e if your code 2.7 you WANT `f-strings` usage to be caught by the linter and a syntax-error displayed. It is also noted in a number of places in the docs, such as in [this *quickstart* note][1].

The discussion on what is the right (or wrong) way to determine python version should be done in [SublimeLinter/issues][1] rather than this plugin, I have a hunch that this was heavily discussed before, just by the amount of issues raised here... That does not mean we can't improve the detection to cover more cases, but it's something that really should be discussed with the maintainers of SublimeLinter.

@JaniKarh sure thing, and good luck!

[1]: http://flake8.pycqa.org/en/latest/index.html#quickstart

**Comment by mandeep** [comment]
@asfaltboy 

Here is the sublimelinter settings. I don't use a project file or sublimelinterrc.

```
{
    "user": {
        "debug": true,
        "delay": 0.25,
        "error_color": "D02000",
        "gutter_theme": "Packages/SublimeLinter/gutter-themes/Default/Default.gutter-theme",
        "gutter_theme_excludes": [],
        "lint_mode": "background",
        "linters": {
            "flake8": {
                "@disable": false,
                "args": [],
                "builtins": "",
                "excludes": [],
                "executable": "",
                "ignore": "",
                "jobs": "1",
                "max-complexity": -1,
                "max-line-length": null,
                "select": "",
                "show-code": false
            },
            "gcc": {
                "@disable": false,
                "args": [],
                "excludes": [],
                "extra_flags": "-Wextra -Wformat=2 -Wshadow -Wundef -Wpointer-arith -Wcast-align -Wstrict-prototypes -std=c11"
            },
            "mlint": {
                "@disable": false,
                "args": [],
                "excludes": []
            },
            "rust": {
                "@disable": true,
                "args": [],
                "crate-root": null,
                "excludes": [],
                "use-cargo": true,
                "use-cargo-check": true,
                "use-crate-root": false
            },
            "shellcheck": {
                "@disable": true,
                "args": [],
                "exclude": "",
                "excludes": []
            }
        },
        "mark_style": "squiggly underline",
        "no_column_highlights_line": false,
        "passive_warnings": false,
        "paths": {
            "linux": [],
            "osx": [],
            "windows": []
        },
        "python_paths": {
            "linux": [],
            "osx": [],
            "windows": []
        },
        "rc_search_limit": 3,
        "shell_timeout": 10,
        "show_errors_on_save": false,
        "show_marks_in_minimap": true,
        "syntax_map": {
            "html (django)": "html",
            "html (rails)": "html",
            "html 5": "html",
            "javascript (babel)": "javascript",
            "magicpython": "python",
            "php": "html",
            "python django": "python",
            "pythonimproved": "python"
        },
        "tooltip_fontsize": "1rem",
        "tooltip_theme": "Packages/SublimeLinter/tooltip-themes/Default/Default.tooltip-theme",
        "tooltip_theme_excludes": [],
        "tooltips": false,
        "warning_color": "DDB700",
        "wrap_find": true
    }
}

```

**Comment by asfaltboy** [comment]
Doesn't seem like you customized your settings, that's great. Since you don't specify any locations, SublimeLinter will rely on Sublime's environment when executing `flake8 --version`, with this particular env, it fails. Try my patch above to see exactly which variables are in sublime, and compare these to the ones set in your shell, the difference should be the cause of the error.

**Comment by eblot** [comment]
Hi all,

Is there an **easy**, **simple** way to tell flake8 to revert back to use Python3. I do not mind if it is dirty, I do not really mind which solution is better or worse, I just like to get things to work as it used to be before Package Manager upgraded it, at least till someone comes with the right clean and lean solution.

Thanks in advance.

**Comment by asfaltboy** [comment]
@eblot have you [configured your linter][1] and looking at the troubleshooting docs? It've mentioned the simplest workaround I know of multiple times in this doc.

[1]: https://github.com/SublimeLinter/SublimeLinter-flake8#linter-configuration

**Comment by eblot** [comment]
@asfaltboy I have added

    "@python": 3,
    "paths": {
         "osx": ["/usr/local/opt/python3/bin/python3.5"],
    },

to my Preferences.sublime-settings file,

but /usr/local/bin/flake8 keeps being launched from Python 2.7.10 since the upgrade.

**Comment by kaste** [comment]
Set **@python** to `/usr/local/opt/python3/bin/python` or whatever points to your good python?

**Comment by joeycarter** [comment]
@eblot For a quick and dirty fix, see [my answer](https://github.com/SublimeLinter/SublimeLinter-flake8/issues/56#issuecomment-338571698) above.

First open `linter.py`:

    PackageResourceViewer: Open Resource > SublimeLinter-flake8 > linter.py

Next, replace,

```python
cmd = ('flake8@python', '*', '-')
```

with,

```python
cmd = ('flake8', '*', '-')
```

and that should do it.

**Comment by eblot** [comment]
@joeycarter Thanks a lot, works like a charm. You save my day :-)

**Comment by asfaltboy** [comment]
@eblot Glad you resolved your issue 😄  ! As to your settings, I believe it will also work if you change:

      "osx": ["/usr/local/opt/python3/bin/python3.5"],

to

      "osx": ["/usr/local/opt/python3/bin"],

**Comment by kylebebak** [comment]
@groteworld 
This is still broken after the update to SublimeLinter 3.10.3, which was [supposedly going to fix the issue](https://github.com/SublimeLinter/SublimeLinter3/issues/641).

I don't understand the issue, but I do know that @joeycarter 's fix of modifying `linter.py` worked. Perhaps you guys ought to release a version that uses `cmd = ('flake8', '*', '-')` instead of `cmd = ('flake8@python', '*', '-')`, i.e. use the quick and dirty fix.

**Comment by kaste** [comment]
I have my PR #57 ready, which needs some testing folks.

**Comment by groteworld** [comment]
This has been closed in #57
